File: web/js/sim/progress.js
// Track-relative localization: arc-length s, signed lateral offset, tangent.
// MUST match racing/sim/progress.py.

import { wrapPi } from "./car.js";

const SEARCH_WINDOW = 12; // centerline indices searched around the hint

export class ProgressTracker {
  constructor(track) {
    this.center = track.centerline; // [[x,z], ...]
    this.sTable = track.s_table;
    this.lapLength = track.lap_length;
    this.n = this.center.length;
  }

  // Project (x,z) onto the centerline near hintIndex (null = global search).
  // Returns { s, lateral, tangentAngle, index }.
  locate(x, z, hintIndex = null) {
    const { center, sTable, lapLength, n } = this;
    let i0;
    let i1;
    if (hintIndex === null) {
      i0 = 0;
      i1 = n - 1;
    } else {
      i0 = hintIndex - SEARCH_WINDOW;
      i1 = hintIndex + SEARCH_WINDOW;
    }
    let best = Infinity;
    let bestIdx = 0;
    let bestT = 0.0;
    for (let k = i0; k <= i1; k++) {
      const i = ((k % n) + n) % n;
      const a = center[i];
      const b = center[(i + 1) % n];
      const dx = b[0] - a[0];
      const dz = b[1] - a[1];
      const len2 = dx * dx + dz * dz;
      let t = 0.0;
      if (len2 > 1e-12) {
        t = ((x - a[0]) * dx + (z - a[1]) * dz) / len2;
        if (t < 0.0) t = 0.0;
        else if (t > 1.0) t = 1.0;
      }
      const cx = a[0] + dx * t;
      const cz = a[1] + dz * t;
      const ddx = x - cx;
      const ddz = z - cz;
      const d2 = ddx * ddx + ddz * ddz;
      if (d2 < best) {
        best = d2;
        bestIdx = i;
        bestT = t;
      }
    }
    const a = center[bestIdx];
    const b = center[(bestIdx + 1) % n];
    const dx = b[0] - a[0];
    const dz = b[1] - a[1];
    const segLen = Math.sqrt(dx * dx + dz * dz);
    const tx = dx / segLen;
    const tz = dz / segLen;
    const cx = a[0] + dx * bestT;
    const cz = a[1] + dz * bestT;
    // Signed lateral: positive on the left of the tangent.
    const lateral = tx * (z - cz) - tz * (x - cx);
    let s = sTable[bestIdx] + segLen * bestT;
    if (s >= lapLength) {
      s -= lapLength;
    }
    return { s, lateral, tangentAngle: Math.atan2(tz, tx), index: bestIdx };
  }

  // Shortest signed arc distance from s0 to s1 around the lap.
  deltaS(s0, s1) {
    let d = s1 - s0;
    if (d > this.lapLength / 2.0) {
      d -= this.lapLength;
    } else if (d < -this.lapLength / 2.0) {
      d += this.lapLength;
    }
    return d;
  }

  headingError(heading, tangentAngle) {
    return wrapPi(heading - tangentAngle);
  }
}

File: racing/sim/progress.py
"""Track-relative localization: arc-length s, signed lateral offset, tangent.

MUST match web/js/sim/progress.js line by line (parity-tested).
"""

import math
from typing import NamedTuple

from racing.sim.car import wrap_pi

SEARCH_WINDOW = 12  # centerline indices searched around the hint


class Location(NamedTuple):
    s: float
    lateral: float
    tangent_angle: float
    index: int


class ProgressTracker:
    def __init__(self, track: dict):
        self.center = track["centerline"]
        self.s_table = track["s_table"]
        self.lap_length = track["lap_length"]
        self.n = len(self.center)

    def locate(self, x, z, hint_index=None) -> Location:
        center = self.center
        n = self.n
        if hint_index is None:
            i0, i1 = 0, n - 1
        else:
            i0, i1 = hint_index - SEARCH_WINDOW, hint_index + SEARCH_WINDOW
        best = math.inf
        best_idx = 0
        best_t = 0.0
        for k in range(i0, i1 + 1):
            i = k % n
            a = center[i]
            b = center[(i + 1) % n]
            dx = b[0] - a[0]
            dz = b[1] - a[1]
            len2 = dx * dx + dz * dz
            t = 0.0
            if len2 > 1e-12:
                t = ((x - a[0]) * dx + (z - a[1]) * dz) / len2
                if t < 0.0:
                    t = 0.0
                elif t > 1.0:
                    t = 1.0
            cx = a[0] + dx * t
            cz = a[1] + dz * t
            ddx = x - cx
            ddz = z - cz
            d2 = ddx * ddx + ddz * ddz
            if d2 < best:
                best = d2
                best_idx = i
                best_t = t
        a = center[best_idx]
        b = center[(best_idx + 1) % n]
        dx = b[0] - a[0]
        dz = b[1] - a[1]
        seg_len = math.sqrt(dx * dx + dz * dz)
        tx = dx / seg_len
        tz = dz / seg_len
        cx = a[0] + dx * best_t
        cz = a[1] + dz * best_t
        # Signed lateral: positive on the left of the tangent.
        lateral = tx * (z - cz) - tz * (x - cx)
        s = self.s_table[best_idx] + seg_len * best_t
        if s >= self.lap_length:
            s -= self.lap_length
        return Location(s, lateral, math.atan2(tz, tx), best_idx)

    def delta_s(self, s0, s1):
        """Shortest signed arc distance from s0 to s1 around the lap."""
        d = s1 - s0
        if d > self.lap_length / 2.0:
            d -= self.lap_length
        elif d < -self.lap_length / 2.0:
            d += self.lap_length
        return d

    def heading_error(self, heading, tangent_angle):
        return wrap_pi(heading - tangent_angle)
